hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/placement/algorithms/SCMContainerPlacementRackScatter.java

```diff
@@ -0,0 +1,332 @@
+    List<Node> chosenNodes = new ArrayList<>();
+    List<Node> unavailableNodes = new ArrayList<>();
+    Set<Node> skippedRacks = new HashSet<>();
+    if (excludedNodes != null) {
+      unavailableNodes.addAll(excludedNodes);
+    }
+
+    // If the result doesn't change after retryCount, we return with exception
+    int retryCount = 0;
+    while (nodesRequired > 0) {
+      if (retryCount > OUTER_LOOP_MAX_RETRY) {
+        throw new SCMException("No satisfied datanode to meet the" +
+            " excludedNodes and affinityNode constrains.", null);
+      }
+      int chosenListSize = chosenNodes.size();
+
+      // Refill toChooseRacks, we put skippedRacks in front of toChooseRacks
+      // for a even distribution
+      if (toChooseRacks.size() == 0) {
+        toChooseRacks.addAll(racks);
+        if (!skippedRacks.isEmpty()) {
+          toChooseRacks.removeAll(skippedRacks);
+          toChooseRacks.addAll(0, skippedRacks);
+          skippedRacks.clear();
+        }
+      }
+
+      if (mutableFavoredNodes.size() > 0) {
+        List<Node> chosenFavoredNodesInForLoop = new ArrayList<>();
+        for (DatanodeDetails favoredNode : mutableFavoredNodes) {
+          Node curRack = getRackOfDatanodeDetails(favoredNode);
+          if (toChooseRacks.contains(curRack)) {
+            chosenNodes.add(favoredNode);
+            toChooseRacks.remove(curRack);
+            chosenFavoredNodesInForLoop.add(favoredNode);
+            unavailableNodes.add(favoredNode);
+            nodesRequired--;
+            if (nodesRequired == 0) {
+              break;
+            }
+          }
+        }
+        mutableFavoredNodes.removeAll(chosenFavoredNodesInForLoop);
+      }
+
+      // If satisfied by favored nodes, return then.
+      if (nodesRequired == 0) {
+        break;
+      }
+
+      List<Node> chosenRacksInForLoop = new ArrayList<>();
+      for (Node rack : toChooseRacks) {
+        List<Node> nodes = ((InnerNode)rack).getNodes(2);
+        if (nodes.size() > 0) {
+          Node affinityNode = nodes.get(0);
+          Node node = chooseNode(unavailableNodes, affinityNode,
+              metadataSizeRequired, dataSizeRequired);
+          if (node != null) {
+            chosenNodes.add(node);
+            mutableFavoredNodes.remove(node);
+            unavailableNodes.add(node);
+            nodesRequired--;
+            if (nodesRequired == 0) {
+              break;
+            }
+          }
+        } else {
+          // Store the skipped racks to check them first in next outer loop
+          skippedRacks.add(rack);
+        }
+        chosenRacksInForLoop.add(rack);
+      }
+      toChooseRacks.removeAll(chosenRacksInForLoop);
+      // If chosenNodes not changed, increase the retryCount
+      if (chosenListSize == chosenNodes.size()) {
+        retryCount++;
+      } else {
+        // If chosenNodes changed, reset the retryCount
+        retryCount = 0;
+      }
+    }
+
```

Comment:
I wonder if we should log a warning if we return a list of nodes where the number of racks is less than the value returned from the `requiredRacks()` method? That would indicate the cluster has enough racks, but for some reason, we have not been able to use enough racks to meet the requirement.

We might need to keep a set of racks select as each node is picked to do this efficiently.